### ISSUE TITLE
PLR1704: does not flag function parameter overwritten by plain assignment

### ISSUE THREAD

**Issue by acampove** [astral-sh/ruff#25228]
## Description

`PLR1704` currently only flags function parameters redefined via `for`, `with`, or `try` statement targets. It does not flag parameters overwritten by a plain assignment inside the function body, even though this is equally confusing and error-prone.

## Example

The following code is not flagged by `PLR1704`:

```python
def process(name: str) -> None:
        name = f"prefix_{item}"  # not flagged
        for x in [1,2]:
            name = 'x' # not flagged either
```

The parameter `name` is silently overwritten, losing the original argument. The caller's intent is discarded with no warning.

## Expected behaviour

Ruff should emit `PLR1704` (or a related rule) when a function parameter is overwritten by a plain assignment anywhere in the function body.

## Actual behaviour

No diagnostic is emitted.

**Comment by ntBre** [comment]
This seems very intentional based on [our documentation](https://docs.astral.sh/ruff/rules/redefined-argument-from-local/#what-it-does) for the rule:

> Checks for variables defined in for, try, with statements that redefine function parameters.

and also that from [pylint](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/redefined-argument-from-local.html):

> Used when a local name is redefining an argument, which might suggest a potential error. This is taken in account only for a handful of name binding operations, such as for iteration, with statement assignment and exception handler assignment.

I didn't find any previous discussion to this effect, but I think this is to support the common pattern of intentionally shadowing a variable, for example:

```py
def process(path: str | Path):
    path = Path(path)
    ...
```

I think applying the rule to any shadowing would be very prone to false positives, even if it does sometimes catch an error.

**Comment by acampove** [comment]
@ntBre 

Variables should not be randomly shapeshifting like that. That' sort of stuff will lead to bugs that are hard to find. I would do:

```python
def process(val : str | Path):
    '''
    Parameters
    -------------------------
    val: Path to file to be processed
    '''
    path = Path(val)

    ...
```

given that what you are giving to `process` is not really a path, but an object that can be transformed into a path. With proper tooling, the user would easily know, from anywhere, what `val` is meant to be through the docstrings

<img width="752" height="758" alt="Image" src="https://github.com/user-attachments/assets/963a41c2-3c13-4b7e-953d-199ea6efe863" />

This would also encourage users to actually use docstrings.
